### PR TITLE
schema: add @bracket.slope in att.tuplet.vis

### LINKED ISSUE #1195: tuplet@slope
It would be useful to add `@slope` to `<tuplet>` and `<tupletSpan>` to allow for explicit horizontal tuplet brackets.  This would be similar to the existing `beam@slope` and `beamSpan@slope`.

See https://github.com/rism-digital/verovio/pull/3442 for discussion.

### PR DESCRIPTION
**Pull request by pe-ro**
To avoid duplicate attribute definitions, att.tuplet.vis is referenced by att.tupletSpan.vis.

Fixes #1195

### CODE REVIEW

File: source/modules/MEI.visual.xml [hunk 1]
```diff
@@ -1933,6 +1933,13 @@
           <rng:ref name="data.STAFFREL.basic"/>
         </datatype>
       </attDef>
+      <attDef ident="bracket.slope" usage="opt">
+         <desc xml:lang="en">Records the slope of the bracket independently of the slope of the beam (if any)
+           encompassing the events of the tuplet.</desc>
```
Comment (craigsapp) [inline source/modules/MEI.visual.xml]: It might be good to explicitly state the units (degrees presumably).

**Review by craigsapp — DISMISSED** [review]

File: source/modules/MEI.visual.xml (hunk 1 above)
Comment (rettinghaus) [reply source/modules/MEI.visual.xml]: refers to #648

**Review by rettinghaus — DISMISSED** [review]

File: source/modules/MEI.visual.xml (hunk 1 above)
Comment (pe-ro) [reply source/modules/MEI.visual.xml]: Degrees sounds good. Other thoughts, @lpugin, @rettinghaus, @kepper, others?

Comment (craigsapp) [reply source/modules/MEI.visual.xml]: The problem I have in #648, is that a non-zero degree value is not overly useful since it would only apply to one specific layout (that was specifically for beams, but also applies to tuplet brackets and brackets in general).   

A better unit would be the vertical offset of the endpoints, which remains constant over all layout spacings (more or less, sine beams may be flattened when the beam angle is too high).

In other words, perhaps something better than degrees would be in diatonic steps (MEI units).   0 means horizontal, 2 = move up by a staff line from the starting point to the ending point.

Comment (pe-ro) [reply source/modules/MEI.visual.xml]: Is this description better --

Records the slope of a bracket in degrees of deviation from horizontal. Positive values indicate upward slope, negative values measure slope downward, and a zero value indicates the bracket is horizontal.  The value here is independent of the slope of the beam (if any) encompassing the events of the tuplet.

Comment (rettinghaus) [reply source/modules/MEI.visual.xml]: Degrees is problematic, because the angle depends on the spacing of the music (as seen in #648). Also you probably would want to limit it then (-90 to 90)? Slope itself is ∆y/∆x, and thus also dependent on spacing. We could define it in MEI as the difference of the distances between the endpoints relative to the staff lines. "3" would mean the end of the bracket is 3 vus higher than its start (at a horizontal staff), "-2.5" it would go down two and a half units.

Comment (pe-ro) [reply source/modules/MEI.visual.xml]: Ok,  how about modeling `@bracket.slope` on `@vo`:

```
<attDef ident="slope" usage="opt">
  <desc xml:lang="en">Records the slope of a bracket in terms of staff 
interline distance; that is, in units of 1/2 the distance between adjacent staff 
lines. Positive values indicate upward slope, negative values measure slope 
downward, and a zero value indicates the bracket is horizontal. The value here 
is independent of the slope of the beam (if any) encompassing the events of 
the tuplet.</desc>
  <datatype>
    <rng:ref name="data.MEASUREMENTSIGNED"/>
  </datatype>
</attDef>
```
To keep the description short, the text starting with "Positive values indicate ..." can go in the `<remarks>` section.

Comment (ahankinson) [reply source/modules/MEI.visual.xml]: Would `beam@slope` also change accordingly? Would be annoying to have two attributes with the same name and different types of values.

**Comment by ahankinson** [comment]
Given that this is new functionality I would suggest holding off merging this until after 5.0.

File: source/modules/MEI.visual.xml (hunk 1 above)
Comment (craigsapp) [reply source/modules/MEI.visual.xml]: > Ok, how about modeling @bracket.slope on `@vo`

That sounds good since a degree value will not be useful for dynamic layout control.

One thought is that `beam@slope` and similar could be changed to `@vo2` in a similar manner to `@tstamp` and `@tstamp2`:  `@vo2` would be the vertical offset of the endpoint compared to the starting point's vertical placement.   This would be clearer than using the name "slope" since that mathematically implies units of degrees.

Slur currently has `@startvo` and `@vo`, so perhaps change these to `@vo` and `@vo2`.  This would match `@tstamp` attribute naming system as well as to `@x`/`@x2` and `@y`/`@y2`.

Comment (rettinghaus) [reply source/modules/MEI.visual.xml]: @craigsapp please stop mixing things up again. `vo` is a vertical offset, accompanied by `@startvo` and `@endvo`. 
@pe-ro yes, please change beam@slope.

**Comment by rettinghaus** [comment]
@ahankinson actually #648 is open for over three years. It would be nice to close this now with 5.0.

**Comment by ahankinson** [comment]
@rettinghaus an open issue doesn’t cause problems, but a hasty solution committed in a rush might, particularly if the conversation involves new ways of measurement and isn’t unified across the spec.

**Comment by rettinghaus** [comment]
@ahankinson generally I agree with you. But the older issue pointed to the problem, which is, that the values in `@slope` didn't have a particular _meaning_, making it basically unusable. Changing it to `data.MEASUREMENTSIGNED` would only fill it with meaning without making existing encodings invalid. So this seems to me not to be a "hasty solution committed in a rush". 
But as you pointed out (and I emphasized) `@bracket.slope` and `beam@slope` need to use the same `data type`.

**Comment by pe-ro** [comment]
I assigned this to the 5.x milestone, but I'd be happy to move it forward to 5.0.

**Comment by pe-ro** [comment]
PR updated.

**Comment by craigsapp** [comment]
> I assigned this to the 5.x milestone, but I'd be happy to move it forward to 5.0.

I don't see a need to rush.  It only became interesting due to refinements in verovio layout (and such refinements can go on forever :-)

File: source/modules/MEI.cmn.xml [hunk 2]
```diff
@@ -291,9 +291,12 @@
         </datatype>
       </attDef>
       <attDef ident="slope" usage="opt">
-        <desc xml:lang="en">Records the slope of the beam.</desc>
+        <desc xml:lang="en">Records the slope of a beam; e.g., in terms of staff interline distance; 
+          that is, in units of 1/2 the distance between adjacent staff lines. Positive values indicate 
+          upward slope, negative values measure slope downward, and a zero value indicates the beam
```
Comment (ahankinson) [inline source/modules/MEI.cmn.xml]:
```suggestion
          upward slope, negative values indicate slope downward, and a zero value indicates the beam 
```

File: source/modules/MEI.visual.xml [hunk 3]
```diff
@@ -238,9 +238,12 @@
         </valList>
       </attDef>
       <attDef ident="beam.slope" usage="opt">
-        <desc xml:lang="en">Captures beam slope.</desc>
+        <desc xml:lang="en">Records the slope of a beam; e.g., in terms of staff interline distance; 
+          that is, in units of 1/2 the distance between adjacent staff lines. Positive values indicate 
+          upward slope, negative values measure slope downward, and a zero value indicates the beam
```
Comment (ahankinson) [inline source/modules/MEI.visual.xml]:
```suggestion
          upward slope, negative values indicate slope downward, and a zero value indicates the beam 
```

File: source/modules/MEI.visual.xml [hunk 4]
```diff
@@ -1933,6 +1936,15 @@
           <rng:ref name="data.STAFFREL.basic"/>
         </datatype>
       </attDef>
+      <attDef ident="bracket.slope" usage="opt">
+         <desc xml:lang="en">Records the slope of a bracket; e.g., in terms of staff interline distance; 
+          that is, in units of 1/2 the distance between adjacent staff lines. Positive values indicate 
+          upward slope, negative values measure slope downward, and a zero value indicates the bracket
```
Comment (ahankinson) [inline source/modules/MEI.visual.xml]:
```suggestion
          upward slope, negative values indicate slope downward, and a zero value indicates the bracket 
```

File: source/modules/MEI.cmn.xml [hunk 5]
```diff
@@ -291,9 +291,12 @@
         </datatype>
       </attDef>
       <attDef ident="slope" usage="opt">
-        <desc xml:lang="en">Records the slope of the beam.</desc>
+        <desc xml:lang="en">Records the slope of a beam; e.g., in terms of staff interline distance;
```
Comment (ahankinson) [inline source/modules/MEI.cmn.xml]: We should refer to these as virtual units (`vu`) instead of re-explaining what a VU is in MEI.

**Review by ahankinson — CHANGES_REQUESTED** [review]

**Comment by pe-ro** [comment]
I think I incorporated @ahankinson's changes, but along the way I managed to create an unnecessary branch I don't know how to remove.

File: source/modules/MEI.cmn.xml [hunk 6]
```diff
@@ -291,9 +291,11 @@
         </datatype>
       </attDef>
       <attDef ident="slope" usage="opt">
-        <desc xml:lang="en">Records the slope of the beam.</desc>
+        <desc xml:lang="en">Records the slope of a beam; e.g., in MEI virtual units. Positive values indicate
```
Comment (ahankinson) [inline source/modules/MEI.cmn.xml:306]: Is it really "e.g., in MEI virtual units" or just "Records the slope of a beam in MEI virtual units." ? If the latter, this should be changed in the other descriptions as well.

Comment (rettinghaus) [reply source/modules/MEI.cmn.xml:306]: `data.MEASUREMENTSIGNED` allows all sorts of units.

Comment (ahankinson) [reply source/modules/MEI.cmn.xml:306]: Hm. Right. OK

**Review by ahankinson — APPROVED** [review]
The `<abbr>` elves might make a visit, but it looks good to me!

**Comment by bwbohl** [comment]
Generally I do agree but would love to see some examples and documentation added to the guidelines ;-)

**Comment by musicEnfanthen** [comment]
Still needs work, moving to 6.

**Comment by bwbohl** [comment]
I think adding some minimal mention of bracket.slope to the guidelines would do the job of a minimum documentation

**Comment by bwbohl** [comment]
@pe-ro, am I assuming correctly that in the context of this pull request, the following branch can be removed: https://github.com/music-encoding/music-encoding/tree/pe-ro-patch-1?

**Comment by pe-ro** [comment]
@bwbohl , you're correct

**Review by rettinghaus — APPROVED** [review]
